zebra/rib.h

```diff
@@ -181,48 +181,35 @@ typedef struct rib_dest_t_
- * top-level chain. As nexthop is in the top level chain, set recursing
- * to 0.
+ * Initialization: Set `nexthop' to the head of the top-level chain.
  *
  * Iteration check: Check that the `nexthop' pointer is not NULL.
  *
  * Iteration step: This is the tricky part. Check if `nexthop' has
- * NEXTHOP_FLAG_RECURSIVE set. If yes, this implies that `nexthop' is in
- * the top level chain and has at least one nexthop attached to
- * `nexthop->resolved'. As we want to descend into `nexthop->resolved',
- * set `recursing' to 1 and set `nexthop' to `nexthop->resolved'.
- * `tnexthop' is left alone in that case so we can remember which nexthop
- * in the top level chain we are currently handling.
+ * NEXTHOP_FLAG_RECURSIVE set. If yes, this implies that `nexthop' has
+ * at least one nexthop attached to `nexthop->resolved', which will be
+ * the next one.
  *
  * If NEXTHOP_FLAG_RECURSIVE is not set, `nexthop' will progress in its
- * current chain. If we are recursing, `nexthop' will be set to
- * `nexthop->next' and `tnexthop' will be left alone. If we are not
- * recursing, both `tnexthop' and `nexthop' will be set to `nexthop->next'
- * as we are progressing in the top level chain.
- *   If we encounter `nexthop->next == NULL', we will clear the `recursing'
- * flag as we arived either at the end of the resolved chain or at the end
- * of the top level chain. In both cases, we set `tnexthop' and `nexthop'
- * to `tnexthop->next', progressing to the next position in the top-level
- * chain and possibly to its end marked by NULL.
+ * current chain. In case its current chain end is reached, it will try
+ * to get up to the previous recursion level and progress there. Whenever
+ * a step forward in a chain is done, recursion will be checked again.
+ * In a nustshell, it's equivalent to a pre-traversal order assuming that
+ * left branch is 'resolved' and right branch is 'next':
+ * https://en.wikipedia.org/wiki/Tree_traversal#/media/File:Sorted_binary_tree_preorder.svg
  */
-#define ALL_NEXTHOPS_RO(head, nexthop, tnexthop, recursing) \
-  (tnexthop) = (nexthop) = (head), (recursing) = 0; \
-  (nexthop); \
-  (nexthop) = CHECK_FLAG((nexthop)->flags, NEXTHOP_FLAG_RECURSIVE) \
-    ? (((recursing) = 1), (nexthop)->resolved) \
-    : ((nexthop)->next ? ((recursing) ? (nexthop)->next \
-                                      : ((tnexthop) = (nexthop)->next)) \
-                       : (((recursing) = 0),((tnexthop) = (tnexthop)->next)))
+#define ALL_NEXTHOPS(head, nexthop)                              \
+  (nexthop) = (head);                                               \
+  (nexthop);                                                        \
+  (nexthop) = CHECK_FLAG((nexthop)->flags, NEXTHOP_FLAG_RECURSIVE)  \
+    ? ((nexthop)->resolved)                                         \
+       : ((nexthop)->next ? (nexthop)->next                         \
+          : ((nexthop)->rparent ? (nexthop)->rparent->next : NULL))
```

Comment:
Assume you have the following nexthop structure:

<pre>NH1 -> NH2
 |
 V
RNH1A
 |
 V
RNH1B</pre>

Wouldn't the iteration go wrong by returning only `NH1`, `RNH1A`, `RNH1B`, because `RNH1B` doesn't have `NEXTHOP_FLAG_RECURSIVE` set, has `->next == NULL` and it's `->rparent` pointing to `RNH1A` so that `->rparent->next` is `NULL` for it?
My expectation for correct behavior would be to get `NH1`, `RNH1A`, `RNH1B`, `NH2` from the iteration. The trouble seems to me that it only goes upwards one level while it needs to go upwards multiple levels: `rparent->next` if this exists, otherwise `rparent->rparent->next`, `rparent->rparent->rparent->next` and so on as long as `rparent->...->rparent != NULL`.
(Indeed, when I reproduced this nexthop structure in `test_nexthop_iter` it triggered an assertion exactly because `NH2` is missing. Looking into this I realized that the PRNG I contributed for testing back in 2013 was actually bulls**t. Once I replaced it with a sensible one, it actually triggered the assertion in the PRNG stage. I will provide an update to the PRNG in another pull request)

As a side note: I am not that fond of the indentation here. At least for me it is much more readable if the `:` aligns with the matching `?` when the expression is split across multiple lines. Otherwise I can easily mixup which belongs to which.